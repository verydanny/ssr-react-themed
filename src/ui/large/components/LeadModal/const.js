export const LEAD_MODAL_ID = 'Modal/LeadModal'
export const THANK_YOU_MODAL_ID = 'Modal/ThankYouModal'
export const LEAD_MODAL_PROPS = [
  'listingId',
  'endecaId',
  'propertyLabel',
  'desktopPhone',
  'tplSource',
  'isPaid',
  'revenue',
]
