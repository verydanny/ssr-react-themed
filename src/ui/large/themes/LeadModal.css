@value ( RentBlue, UniversalGrey, Gray1 ) from 'ui/shared/themes/colors/colors.css';

.LeadModal {
}

.LeadModal .Modal {
  width: 374px;
  height: 604px;
  /* Prevent modal from going off-screen at the top for small window height */
  max-height: 100vh;
  /* Override position absolute which is being set elsewhere */
  position: static;

  /* In case the modal is taller than window size */
  overflow-y: auto;
}

.LeadModal .Modal::-webkit-scrollbar {
  width: 0px;
  background: transparent;
}

.LeadModal .ModalBody {
  padding: 16px;
  min-height: calc(100% - 32px);
  background: #fff;
}

.LeadModal .ModalBody:after {
  content: "";
  display: table;
  clear: both;
}

.LeadModal .Modal-open {
  display: block;
}

.LeadModal .Model-close {
  display: none;
}

.LeadModal .CloseButton {
  float: right;
  font-size: 20px;
  cursor: pointer;
}

.LeadModal .CloseButton svg {
  width: 1em;
  height: 1em;
}

.LeadModal .Modal_Header {
  font-weight: 700;
  font-size: 18px;
  line-height: 22px;
  color: UniversalGrey;
  text-decoration: none;
}

.LeadModal .Modal_Header a {
  color: RentBlue;
}

.LeadModal .Modal_SubHeader {
  color: Gray1;
  font-size: 16px;
  font-weight: 700;
  padding-top: 15px;
  padding-bottom: 13px;
}

.LeadModal .Modal_SubHeader_PropertyTitle {
  color: UniversalGrey;
}
