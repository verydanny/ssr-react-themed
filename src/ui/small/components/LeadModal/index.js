export { LEAD_MODAL_ID, LEAD_MODAL_PROPS } from './const'
export LeadModal from './container'
