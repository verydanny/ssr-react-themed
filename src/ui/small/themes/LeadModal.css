@value ( RentBlue, UniversalGrey, Gray1 ) from 'ui/shared/themes/colors/colors.css';

.LeadModal .Modal {
  position: absolute;
  z-index: 10000;
  top: 0;
  width: 100%;
  height: 100%;
  overflow-y: auto;
}

.LeadModal .ModalBody {
  padding: 16px;
  min-height: 100%;
  background: #fff;
}

.LeadModal .ModalBody:after {
  content: "";
  display: table;
  clear: both;
}

.LeadModal .Overlay {
  position: fixed;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.3);
  z-index:100;
}

.LeadModal .Modal-open {
  display: block;
}

.LeadModal .Model-close {
  display: none;
}

.LeadModal .CloseButton {
  float: right;
  font-size: 20px;
}

.LeadModal .CloseButton svg {
  width: 1em;
  height: 1em;
}

.LeadModal .Modal_Header {
  font-weight: 700;
  font-size: 18px;
  line-height: 22px;
  color: UniversalGrey;
  text-decoration: none;
}

.LeadModal .Modal_Header a {
  color: RentBlue;
}

.LeadModal .Modal_SubHeader {
  color: Gray1;
  font-size: 16px;
  font-weight: 700;
  padding-top: 15px;
  padding-bottom: 13px;
}

.LeadModal .Modal_SubHeader_PropertyTitle {
  color: UniversalGrey;
}
