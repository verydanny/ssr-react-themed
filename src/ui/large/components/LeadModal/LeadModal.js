import React, { Component, Fragment } from 'react'
import PropTypes, { instanceOf } from 'prop-types'
import autobind from 'autobind-decorator'
import { Tagging } from '@rentpath/react-ui-tracking'
import { themed } from 'react-themed-too'
import { LeadModal as LeadModalUI } from '@rentpath/react-ui-rent'
import { Cookies } from 'react-cookie'
import { getLeadData } from 'lib/cookies/leadCookie'
import isoToLocaleDate from 'lib/utils/isoToLocaleDate'
import internalCookieBasedFields from 'lib/lead/cookieBasedFields'
import get from 'lodash/fp/get'
import camelCase from 'lodash/camelCase'
import formatPhone from 'lib/utils/formatPhone'
import { Icon } from 'ui/shared/components/Icon'
import { close } from 'ui/shared/components/Icon/svgs/global'
import Spinner from 'ui/shared/components/Spinner'
import { desktopRevenueMultiplier } from 'lib/lead/revenue'
import {
  ValidatedName,
  ValidatedEmail,
  ValidatedPhone,
  ValidatedMoveInDate,
} from './fields'

const DEFAULT_FIELD_VALUES = {
  message: '',
  name: '',
  email: '',
  phone: '',
  opt_in_newsletter: 0,
}

@themed('*')
export default class LeadModal extends Component {

  static propTypes = {
    cookies: instanceOf(Cookies).isRequired,
    isOpen: PropTypes.bool,
    leadData: PropTypes.object.isRequired,
    listingId: PropTypes.string.isRequired,
    listHubId: PropTypes.string,
    endecaId: PropTypes.string,
    desktopPhone: PropTypes.string,
    isPaid: PropTypes.bool,
    onClose: PropTypes.func.isRequired,
    propertyLabel: PropTypes.string.isRequired,
    revenue: PropTypes.string,
    submitLead: PropTypes.func,
    submitLeadSuccess: PropTypes.func,
    submitLeadFailure: PropTypes.func,
    theme: PropTypes.object,
    tagSection: PropTypes.string,
    tplSource: PropTypes.string,
    isCountryUS: PropTypes.bool,
  }

  static defaultProps = {
    theme: {},
    tagSection: 'lead_submission_form',
    isCountryUS: false,
  }

  constructor(props) {
    super(props)

    const { cookies, isCountryUS } = this.props

    const cookieData = getLeadData(cookies)

    const optInNewsletter = isCountryUS ? 1 : 0

    this.state = {
      isFetching: false,
      fieldStatus: {},
      leadData: {
        ...DEFAULT_FIELD_VALUES,
        opt_in_newsletter: optInNewsletter,
        ...cookieData,
      },
    }
  }

  get desktopRevenue() {
    return desktopRevenueMultiplier(
      this.props.revenue,
      this.cookieBasedFields.campaignId
    )
  }

  get header() {
    const { theme, desktopPhone, isPaid } = this.props

    const phoneStr = this._phone || (this._phone = formatPhone(desktopPhone))
    const phone = phoneStr && isPaid ?
      (<a
        href={`tel:${phoneStr}`}
        className={theme.Modal_Header}
        data-tag_item="phone_number_link"
        data-tid="lead_form_phone_header"
      >
        {phoneStr}
      </a>)
      : null
    return (
      <span data-tid="contact-property-text">Contact Property {phone}</span>
    )
  }

  get subHeader() {
    const {
      propertyLabel,
      theme,
    } = this.props

    return (
      <span data-tid="to-text">
        To:{' '}
        <span className={theme.Modal_SubHeader_PropertyTitle} data-tid="lead-form-listing-name">
          {propertyLabel}
        </span>
      </span>
    )
  }

  get cookieBasedFields() {
    const { cookies, endecaId } = this.props
    return internalCookieBasedFields(cookies, endecaId)
  }

  get messagePlaceholder() {
    const {
      propertyLabel,
    } = this.props

    return [
      `Hello, I'm interested in ${propertyLabel}.`,
      'Please send me current availability and additional details.',
      'Thanks!',
    ].join(' ')
  }

  get submitContent() {
    const { theme } = this.props

    if (this.state.isFetching) {
      return <Spinner className={theme.Spinner} />
    }

    return 'Send'
  }

  get leadFields() {
    const { theme, isCountryUS } = this.props
    const { fieldStatus, leadData } = this.state

    const allFields = [
      {
        name: 'message',
        'data-tag_item': 'text',
        placeholder: this.messagePlaceholder,
        value: leadData.message,
        onChange: this.handleInputChange,
      },
      {
        name: 'name',
        field: ValidatedName,
        'data-tag_item': 'name',
        autoComplete: 'name',
        value: leadData.name,
        onChange: this.handleInputChange,
        ...fieldStatus.name,
      },
      {
        name: 'email',
        field: ValidatedEmail,
        'data-tag_item': 'email',
        autoComplete: 'email',
        value: leadData.email,
        onChange: this.handleInputChange,
        ...fieldStatus.email,
      },
      {
        name: 'phone',
        field: ValidatedPhone,
        placeholder: 'Phone (optional)',
        'data-tag_item': 'phone_number',
        autoComplete: 'tel-national',
        value: leadData.phone,
        onChange: this.handleInputChange,
        ...fieldStatus.phone,
      },
      {
        name: 'move_date',
        field: ValidatedMoveInDate,
        'data-tag_item': 'move_in_date',
        defaultValue: isoToLocaleDate(leadData.move_date),
        ...fieldStatus.moveDate,
      },
      {
        name: 'submit',
        'data-tag_item': 'send_button',
        children: this.submitContent,
        className: theme.ButtonRaised,
      },
      {
        name: 'opt_in_newsletter',
        'data-tag_item': 'checkbox',
        checked: leadData.opt_in_newsletter,
        onChange: this.handleInputChange,
        className: theme.LeadForm_Checkbox,
      },
      { name: 'terms_of_service' },
    ]

    return allFields.filter(field => !(field.name === 'opt_in_newsletter' && isCountryUS))
  }

  @autobind
  setFieldStatus(apiErrors) {
    if (!apiErrors || !apiErrors.length) return

    const checkFields = {
      name: 'Name',
      email: 'Email address',
      phone: 'Phone number',
      moveDate: 'Move in date',
    }

    /* eslint-disable no-param-reassign */
    const fieldStatus = apiErrors.reduce((accum, error) => {
      if (error && error.name && error.message) {
        accum[error.name] = {
          validation: 'invalid',
          validationMessage: `${checkFields[error.name]} ${error.message}`,
        }
      }
      return accum
    }, {})
    /* eslint-enable no-param-reassign */

    Object.keys(checkFields).forEach(input => {
      if (!fieldStatus[input]) {
        fieldStatus[input] = { validation: 'valid' }
      }
    })

    this.setState({ fieldStatus })
  }

  @autobind
  handleInputChange(event) {
    const target = event.target
    const value = target.type === 'checkbox' ? target.checked : target.value

    this.setState({
      leadData: {
        ...this.state.leadData,
        [target.name]: value,
      },
    })
  }

  @autobind
  closeButton({ onClick }) {
    const { theme } = this.props

    return (
      <div
        className={theme.CloseButton}
        onClick={onClick}
        role="button"
        tabIndex={0}
        data-tag_item="x"
      >
        <Icon svgs={{ close }} />
      </div>
    )
  }

  @autobind
  submitLeadResult(result) {
    this.setState({ isFetching: false })

    const errors = get('data.lead.errors')(result)
    const status = get('data.lead.status')(result)

    const {
      submitLeadSuccess,
      submitLeadFailure,
    } = this.props

    switch (status) {
      case 'FIELD_ERROR':
        this.trackLead(
          'error', {
            selection: 'validation',
            validation_error: errors.map(error => error.name),
          }
        )
        this.setFieldStatus(errors)
        break
      case 'ACCEPTED':
        if (submitLeadSuccess) submitLeadSuccess(this.props.listingId)
        this.trackLead(
          'lead_submission', {
            selection: 'email',
            revenue: this.desktopRevenue,
          }
        )
        break
      case 'FAILED':
        if (submitLeadFailure) submitLeadFailure(this.props.listingId)
        break
      default:
        // TODO: Need a generic error message for API failure
    }
  }

  @autobind
  submitLeadError() {
    this.setState({ isFetching: false })
    // TODO: Need a generic error message for connection failure
  }

  trackLead(action, data) {
    window.eventTracker.track(action, data)
  }

  @autobind
  submitLead(event, formData) {
    const {
      leadData,
      listingId,
      endecaId,
      submitLead,
      tplSource,
      isCountryUS,
    } = this.props

    if (this.state.isFetching) return

    const optInNewsletter = isCountryUS || formData.opt_in_newsletter === 'on'

    const mergedData = {
      ...formData,
      message: formData.message || this.messagePlaceholder,
      opt_in_newsletter: optInNewsletter,
      move_date: formData.move_date,
      endecaId,
      listingId,
      tplSource,
      ...this.cookieBasedFields,
      ...leadData,
    }

    const variables = Object.keys(mergedData)
      .filter(k => mergedData[k] !== null)
      .reduce((acc, key) => {
        acc[camelCase(key)] = mergedData[key]
        return acc
      }, {})

    const promise = { resolve: this.submitLeadResult, reject: this.submitLeadError }

    this.setState({ isFetching: true, fieldStatus: {} })
    submitLead(variables, promise)
  }

  render() {
    const {
      isOpen,
      onClose,
      theme,
      listingId,
      listHubId,
      tagSection,
      tplSource,
    } = this.props

    return (
      <Fragment>
        <Tagging
          key="lead_tagging"
          listing_id={listingId}
          listhub_id={listHubId}
          section={tagSection}
          tpl_source={tplSource}
        />
        <div
          className={theme.LeadModal}
        >
          <LeadModalUI
            data-tag_section={tagSection}
            CloseButton={this.closeButton}
            fields={this.leadFields}
            header={this.header}
            isOpen={isOpen}
            onClose={onClose}
            onSubmit={this.submitLead}
            serialize
            subHeader={this.subHeader}
          />
        </div>
      </Fragment>
    )
  }
}
