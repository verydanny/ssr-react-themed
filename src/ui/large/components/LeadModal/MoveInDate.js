import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'
import autobind from 'autobind-decorator'
import { themed } from 'react-themed-too'
import { Field } from '@rentpath/react-ui-core'
import { Icon } from 'ui/shared/components/Icon'
import calendar from 'ui/shared/components/Icon/svgs/calendar.svg'
import { startDate, minDate, maxDate } from 'lib/lead/moveInDate'
import isoToLocaleDate from 'lib/utils/isoToLocaleDate'
import dateToInt from 'lib/utils/dateToInt'
import { format, parse } from 'date-fns'
import { Subject } from 'rxjs/Subject'

const externalOpenHandler = new Subject()

@themed('*')
export default class MoveInDate extends Component {
  static propTypes = {
    theme: PropTypes.object,
    variant: PropTypes.string,
    defaultValue: PropTypes.string,
  }

  static defaultProps = {
    theme: {},
  }

  constructor(props) {
    super(props)

    this.startDate = isoToLocaleDate(startDate())
    this.minDate = minDate()
    this.maxDate = maxDate()

    this.state = {
      date: props.defaultValue || this.startDate,
    }

    this.datePickerStartDate = parse(this.state.date, 'MM/DD/YYYY')
  }

  @autobind
  dateTagging(date) {
    if (window.eventTracker) {
      window.eventTracker.track('click', {
        item: 'move_in_date',
        section: 'lead_submission_form',
        selection: date,
      })
    }
  }

  @autobind
  validateDate(event) {
    // Value used for tagging and local state

    const value = format(event, 'MM/DD/YYYY')

    this.dateTagging(value)

    const min = dateToInt(this.minDate)
    const max = dateToInt(this.maxDate)
    const current = dateToInt(value)

    // Todo - move this out of local state and into Redux
    if (current < min) {
      this.setState({ date: this.minDate })
    } else if (current > max) {
      this.setState({ date: this.maxDate })
    } else {
      this.setState({ date: value })
    }
  }

  openCalendar(e) {
    // TODO: change this to use new handler if needed
    // Once we can redeploy react-ui, we can remove this, and use the
    // new datepicker event handler
    e.nativeEvent.stopImmediatePropagation()
    const parentIsCalender = !!e.target.closest('.react-datepicker')

    if (!parentIsCalender) {
      externalOpenHandler.next()
    }
  }

  render() {
    const { theme, variant, defaultValue, ...filteredProps } = this.props

    return (
      <div
        className={cn(
          theme.MoveInDate_Container,
          variant && theme[`MoveInDate-${variant}`]
        )}
        data-tid="datePickerWrapper"
        role="presentation"
        onClick={this.openCalendar}
      >
        <Icon svgs={{ calendar }} className={theme.MoveInDate_Calendar} />
        <Field
          name="move_date"
          dateFormat="MM/DD/YYYY"
          variant={variant}
          type="datePicker"
          data-tag_item="move_in_date"
          className={theme.MoveInDate_DateField}
          value={this.state.date}
          startDate={this.datePickerStartDate}
          minDate={this.minDate}
          maxDate={this.maxDate}
          onChange={this.validateDate}
          externalOpenHandler={externalOpenHandler}
          {...filteredProps}
        />
      </div>
    )
  }
}
