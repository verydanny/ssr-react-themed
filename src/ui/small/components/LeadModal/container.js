import { connect } from 'react-redux'
import {
  submitLead,
  submitLeadSuccess,
  submitLeadFailure,
} from 'app/store/shared/actions'
import { withCookies } from 'react-cookie'
import pipe from 'lodash/fp/pipe'
import { lead as leadDataSelector } from 'app/selectors/lead'
import { getListHubId } from 'app/selectors/listings'
import { isCountryUS } from 'app/selectors/geoIp'

import LeadModal from './LeadModal'

const mapDispatchToProps = {
  submitLead,
  submitLeadSuccess,
  submitLeadFailure,
}

const mapStateToProps = state => ({
  isCountryUS: isCountryUS(state),
  listHubId: getListHubId(state),
  ...leadDataSelector(state),
})

export default pipe(withCookies, connect(mapStateToProps, mapDispatchToProps))(LeadModal)
