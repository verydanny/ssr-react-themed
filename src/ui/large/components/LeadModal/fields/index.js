import {
  Name,
  Email,
  Phone,
} from '@rentpath/react-ui-rent/lib/LeadForm'
import MoveInDate from '../MoveInDate'
import validatedField from './ValidatedField'

export const ValidatedName = validatedField(Name)

export const ValidatedEmail = validatedField(Email)

export const ValidatedPhone = validatedField(Phone)

export const ValidatedMoveInDate = validatedField(MoveInDate)
