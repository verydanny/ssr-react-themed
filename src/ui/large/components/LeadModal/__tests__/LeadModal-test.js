import React from 'react'
import { render } from 'enzyme'
import { Cookies } from 'react-cookie'
import {
  leadFormHiddenFieldsTest as sharedHiddenFieldsTest,
} from '__helpers__'
import LeadModal from '../LeadModal'

const trackSpy = jest.fn()
window.eventTracker = {
  track: trackSpy,
}

describe('LeadModal', () => {
  describe('hidden fields', sharedHiddenFieldsTest(LeadModal))

  describe('tagging', () => {
    const props = {
      listingId: '111',
      listHubId: '111',
      propertyLabel: 'Test Property Label',
      desktopPhone: '123-456-7890',
      isPaid: true,
      onClose: () => true,
      isOpen: true,
      leadData: {},
      cookies: new Cookies(),
      isCountryUS: false,
    }
    const wrapper = render(<LeadModal {...props} />)

    describe('props.isCountryUS', () => {
      it('hides checkbox when true', () => {
        const component = render(<LeadModal {...props} isCountryUS />)
        expect(
          component.find('[data-tid="lead-form-opt_in_newsletter"]').length
        ).toBe(0)
      })

      it('disabled checkbox when false', () => {
        const component = render(<LeadModal {...props} />)
        expect(
          component.find('[data-tid="lead-form-opt_in_newsletter"]').prop('checked')
        ).toBeFalsy()
      })
    })

    it('tags close button', () => {
      expect(wrapper.find('[data-tag_item="x"]').length)
        .toMatchSnapshot()
    })

    it('tags phone number link', () => {
      expect(wrapper.find('[data-tag_item="phone_number_link"]').length)
       .toMatchSnapshot()
    })

    it('tags text input', () => {
      expect(wrapper.find('[data-tag_item="text"]').length)
        .toMatchSnapshot()
    })

    it('tags name input', () => {
      expect(wrapper.find('[data-tag_item="name"]').length)
        .toMatchSnapshot()
    })

    it('tags phone_number input', () => {
      expect(wrapper.find('[data-tag_item="phone_number"]').length)
        .toMatchSnapshot()
    })

    it('tags lead_submission_form section', () => {
      expect(wrapper.find('[data-tag_section="lead_submission_form"]').length)
        .toMatchSnapshot()
    })
  })
})
