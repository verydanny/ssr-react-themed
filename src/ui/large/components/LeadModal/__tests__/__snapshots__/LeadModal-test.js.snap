// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeadModal tagging tags close button 1`] = `1`;

exports[`LeadModal tagging tags lead_submission_form section 1`] = `1`;

exports[`LeadModal tagging tags name input 1`] = `1`;

exports[`LeadModal tagging tags phone number link 1`] = `1`;

exports[`LeadModal tagging tags phone_number input 1`] = `1`;

exports[`LeadModal tagging tags text input 1`] = `1`;
